homeassistant/config_entries.py

```diff
@@ -1610,34 +1611,48 @@ def __setitem__(self, entry_id: str, entry: ConfigEntry) -> None:
         data[entry_id] = entry
         self._index_entry(entry)
 
+    def check_unique_id(self, entry: ConfigEntry) -> None:
+        """Check config entry unique id.
+
+        For a string unique id (this is the correct case): return
+        For a hashable non string unique id: log warning
+        For a non-hashable unique id: raise error
+        """
+        if (unique_id := entry.unique_id) is None:
+            return
+        if isinstance(unique_id, str):
+            # Unique id should be a string
+            return
+        if isinstance(unique_id, Hashable):  # type: ignore[unreachable]
```

Comment:
I think we should add back the string copy of the unique id for this case so we don't break existing integrations that set invalid unique_ids that are hashable but not string.

See issue https://github.com/home-assistant/core/issues/129522

Reply:
Or something in the middle - as we have had lots of changes from @emontnemery regarding uniqueness of unique ids

Reply:
There's nothing in the code anymore that stringify the hashable unique id (it was removed with this PR).
I'll make a fix for this later

Reply:
That's my point - I don't see where it was removed.
As far as I can see it was only removed for non-hashable...

Reply:
I think you're right. It seems it was previously stored as it came in e.g. `int` etc.
This PR in itself seems to not change this behavior.

Reply:
Erik found that the Supervisor library client returns a UUID instance instead of a string for the uuid attribute of the supervisor discovery info. That's what we see in the issue.

Reply:
Yes, but as Erik wrote me on discord we should update the logging a bit to clarify the type coming in. Maybe now obvious it was an UUID but maybe not in many other cases.

Reply:
Very good!